### PR TITLE
Jonmv/multi range item 2

### PR DESCRIPTION
**Pull request by jonmv**
@baldersheim please review. #19218 for context.

### CODE REVIEW

**Comment by jonmv** [comment]
Need to prune e.g., (1, 1), when [0, 1) is already present. That isn't handled now, as these two do not overlap.

**Review by baldersheim — CHANGES_REQUESTED** [review]
See my comments. Overall this looks good.

But if you want to bundle api and encoding then this will not be usable for a long time.
This introduces 2 indexes with a term and that is a new concept that does not exist in the backend. Adding support for this will take quite some effort.

I would probably serialize as today as a first step. Then they would get the benefit and you would get feedback if the API is suitable.

File: container-search/src/main/java/com/yahoo/prelude/query/MultiRangeItem.java [hunk 1]
```diff
@@ -0,0 +1,297 @@
...
+    }
+
+    @Override
+    void encodeBlueprint(ByteBuffer buffer) {
+        boolean sameIndex = startIndex.equals(endIndex);
+        byte metadata = 0;
+        if (sameIndex)      metadata |= 0b00000001;
+        if (startInclusive) metadata |= 0b00000010;
+        if (endInclusive)   metadata |= 0b00000100;
+        buffer.put(metadata);
```
[the remark this answers is not in the document]
Comment (jonmv) [reply container-search/src/main/java/com/yahoo/prelude/query/MultiRangeItem.java:276]: Added type argument, with support for int, long and double.
Also added dynamic selection of integer compression, which can be extended later.

File: container-search/src/main/java/com/yahoo/prelude/query/MultiRangeItem.java [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+        putString(startIndex, buffer);
+        if ( ! sameIndex) putString(endIndex, buffer);
+    }
+
+    @Override
+    void encodeTerms(ByteBuffer buffer) {
+        for (Range range : sortedRanges()) {
+            buffer.putDouble(range.start.doubleValue());
```
[the remark this answers is not in the document]
Comment (jonmv) [reply container-search/src/main/java/com/yahoo/prelude/query/MultiRangeItem.java]: Done.

File: container-search/src/main/java/com/yahoo/prelude/query/MultiRangeItem.java [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+            }
+        }
+        pruned.add(start == end ? start : new Range(start.start, end.end));
+
+        sorted = true;
+        return ranges = pruned;
+    }
+
+    @Override
+    public void setIndexName(String index) { } // This makes no sense :<
```
[the remark this answers is not in the document]
Comment (jonmv) [reply container-search/src/main/java/com/yahoo/prelude/query/MultiRangeItem.java]: Yes ...

**Comment by jonmv** [comment]
PTAL.

**Review by baldersheim — APPROVED** [review]

**Pull request merged**